torch.compile should auto-functionalize certain mutable ops

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115135
* #115134
* #114956
* __->__ #114955

Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

zou3519 committed:
torch.compile should auto-functionalize certain mutable ops

Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "torch.compile should auto-functionalize certain mutable ops"


Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "torch.compile should auto-functionalize certain mutable ops"


Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "torch.compile should auto-functionalize certain mutable ops"


Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on test/dynamo/test_misc.py:
```diff
@@ -404,6 +405,116 @@ def h(x):
+            orig_args = (x, y, z, n)
+
+            compiled_args = pytree.tree_map_only(torch.Tensor, torch.clone, orig_args)
+            torch.compile(f, backend="aot_eager")(*compiled_args)
```
Just a thought: `aot_eager_decomp_partition` might be a better test backend here: the `aot_eager` backend actually does not try to keep input mutations in the graph (so the other backend will be a little closer to the "default" torch.compile behavior).

We could maybe update `aot_eager` to keep input mutations in the graph, but I thought it would be a nice way to debug compiler issues specific to input mutations if we keep aot_eager as is.

zou3519 replied:
I didn't know about this backend, let me try it out

bdhirsh commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -0,0 +1,192 @@
+auto_functionalized = AutoFunctionalized()
+
+
+def can_auto_functionalize(op: torch._ops.OperatorBase) -> bool:
```
Do you think it's returning False for ops in the ATen namespace? Just so we don't accidentally auto-functionalize them (even though I think in way you use this in `fake_tensor.py` today, that will never happen since you only try to run this during the fallback).

My thought was that this will be silently wrong for metadata-mutation ops - and even once you add op_tests to error on metadata-mutation, it's still valid to have aten ops that mutate metadata.

bdhirsh commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -0,0 +1,192 @@
+) -> Tuple[Tensor, ...]:
+    new_kwargs = dict(**kwargs)
+    result = []
+    for name in mutated_args_names:
```
Do you think it's worth a `warn_once()` here as a perf warning? (counterargument == too noisy, and it's not very actionable)

zou3519 replied:
Probably not for now, as you mentioned that it is not very actionable. Maybe we can do this in the future when the re-inplacing pass exists (and we can log instead of warn so someone staring at the log might be able to surface this to us)

bdhirsh reviewed (approved):
left some comments but lgtm

zou3519 commented on test/dynamo/test_misc.py:
```diff
@@ -404,6 +405,116 @@ def h(x):
+            )
+
+            @torch.library.impl("mylib::foo", "cpu", lib=lib)
+            @torch._dynamo.disable
```
Dynamo interposes on all frames. The Pytorch dispatcher invoking this function is a frame, so Dynamo ends up trying to compile it. Dynamo errors out loudly here because the Python key has been disabled at this point, but it tries to construct FakeTensors. The workaround is to disable Dynamo here, but a longer-term solution is to get Dynamo to reset dispatcher state temporarily.

bdhirsh replied:
> The Pytorch dispatcher invoking this function is a frame, so Dynamo ends up trying to compile it. Dynamo errors out loudly here because the Python key has been disabled at this point, but it tries to construct FakeTensors

hmm. I'm sure this is a noob question, but - this is a python function getting invoked by C++ code (the pytorch dispatcher). I had always just assumed that any python getting called by C++ code was "invisible" to dynamo. If dynamo will indeed go ahead and try to compile this python frame (and fakeify the inputs) even though it was called from C++, then why doesn't e.g. every tensor subclass's `__torch_dispatch__` not also need to be wrappe in a `@dynamo.disable`?

zou3519 replied:
> This is a python function getting invoked by C++ code (the pytorch dispatcher). I had always just assumed that any python getting called by C++ code was "invisible" to dynamo

My understanding is that python functions called by C++ code still get seen by Dynamo.

>  If dynamo will indeed go ahead and try to compile this python frame (and fakeify the inputs) even though it was called from C++, then why doesn't e.g. every tensor subclass's __torch_dispatch__ not also need to be wrappe in a @dynamo.disable?

We execute the`__torch_dispatch__` as a part of AOTAutograd, right? And Dynamo is turned off during AOTAutograd.

eellison commented:
How does this work if there are aliased inputs to an op? In that case, won't cloning before a mutation potentially be incorrect ?

zou3519 commented:
@eellison that's a good test case, I'll add one. We expect this to just work automatically (https://gist.github.com/zou3519/373634e2be660026252969882fa7fe84)

The TL;DR is that when we replace the original input with the new output during Functionalization, if the input to be replaced is a view, we do some heavy-lifting of ensuring that each view gets updated

zou3519 committed:
Update on "torch.compile should auto-functionalize certain mutable ops"


Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "torch.compile should auto-functionalize certain mutable ops"


Users may wish to torch.compile custom ops that mutate their inputs
and return nothing (this is a common class of operators).
torch.compile will automatically support this op without anyone needing
to provide a functionalization kernel for it. Here's how.

Let's say we have a hypothetical mylib::sin_(Tensor(a!) x) -> ()
op. First, when FakeTensor sees this op, it can just return None.
This is the case because custom ops are not allowed to mutate input
metadata, so the FakeTensor rule for one that returns nothing is trivial.

Next, when Python FunctionalTensor sees the op, it will functionalize
it by emitting a call to an auto_functionalize(op, ["x"], {"x": ...})
HOP and replacing the mutated inputs with the outputs of this HOP.
This HOP effectively runs the functional version of the op when
called: it clones inputs that will be mutated, runs the op, and
then returns Tensors with the new values.

In the future we can teach Inductor how to do re-inplacing when it sees
this HOP (like how triton kernels do it) but this isn't urgent (and is
more of a performance problem).

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
